src/frontend/lexer.rs

```diff
@@ -0,0 +1,259 @@
+use itertools::Itertools;
+use std::iter::Peekable;
+
+use super::token::Token;
+
+// -------------------------------------------------------------------------- //
+// Error type                                                                 //
+// -------------------------------------------------------------------------- //
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Error {
+    UnexpectedSymbol(char),
+    UnclosedString,
+}
+
+// -------------------------------------------------------------------------- //
+// Type aliases                                                               //
+// -------------------------------------------------------------------------- //
+
+type TokResult = Result<Token, Error>;
+
+// -------------------------------------------------------------------------- //
+// Lexer struct                                                               //
+// -------------------------------------------------------------------------- //
+
+pub struct Lexer<I: Iterator<Item = char>> {
+    source: Peekable<I>,
+}
+
+impl<I: Iterator<Item = char>> Lexer<I> {
+    fn new(source: Peekable<I>) -> Self {
+        Self { source }
+    }
+
+    // ---------------------------------------------------------------------- //
+    // Primary lexing method                                                  //
+    // ---------------------------------------------------------------------- //
+
+    fn next_token(&mut self) -> Option<TokResult> {
+        use Token::*;
+
+        self.skip_whitespace();
+        let next_char = self.peek()?;
+
+        let next_token = match next_char {
+            '(' => self.accept(LParen),
+            ')' => self.accept(RParen),
+            '\'' => self.accept(Quote),
+            c if Self::starts_identifier(c) => self.parse_identifier(),
+            c if Self::starts_integer(c) => self.parse_integer(),
+            c if Self::starts_string(c) => self.parse_string(),
+            // Unexpected symbol
+            c => Err(Error::UnexpectedSymbol(c.clone())),
+        };
+
+        Some(next_token)
+    }
+
+    // ---------------------------------------------------------------------- //
+    // Whitespace handling                                                    //
+    // ---------------------------------------------------------------------- //
+
+    fn is_whitespace(c: &char) -> bool {
+        c.is_ascii_whitespace()
+    }
+    fn skip_whitespace(&mut self) {
+        self.skip_while(Self::is_whitespace);
+    }
+
+    // ---------------------------------------------------------------------- //
+    // Token specific lexing methods                                          //
+    // ---------------------------------------------------------------------- //
+
+    // Identifier ----------------------------------------------------------- //
```

Comment:
nit, is there a way to subcategorize these programatically using private subclasses (or the Rust equivalent) to reduce reliance on comments?

Reply:
I wouldn't say I'm relying on them. But I have realized that nobody uses anything of the sort nowadays so... I think I'll just not write these in the future :)